pilot/pkg/proxy/envoy/xds/xds.go

```diff
@@ -122,6 +123,26 @@ func NewXDS() *Server {
 	return s
 }
 
+// WaitConfigSync will wait for the memory controller to sync.
+func (s *Server) WaitConfigSync(max time.Duration) bool {
+	// TODO: when adding support for multiple config controllers (matching MCP), make sure the
+	// new stores support reporting sync events on the syncCh, to avoid the sleep loop from MCP.
+	if s.ConfigStoreCache.HasSynced() {
+		return true
+	}
+	maxCh := time.After(max)
+	for {
+		select {
+		case <-s.syncCh:
+			if s.ConfigStoreCache.HasSynced() {
```

Comment:
I don't get why we need these hacks. XDS already has a way to do this. We send all of our DiscoveryRequests then wait until we get DiscoveryResponses for them all. Why do we need to have this logic of creating empty objects?

Reply:
The memory store interface is based on 'set' calls for individual objects. We know when we got all the DiscoveryResponse objects - but memory store doesn't. 

I haven't created a 'MarkSynced' interface because I'm not entirely sure even the current HasSynced is right. We also have the debouncing code and not clear how federation will deal with this - I don't want to jump to a top-level interface until I understand the full problem.